### DISCUSSION TITLE
Absolute paths

### DISCUSSION THREAD

**Discussion by sjg20** [labgrid-project/labgrid#1389] [Q&A]
Why does labgrid use relative paths for imports? When using code inside and outside labgrid, it requires transformation.

Would it not be easier to use absolute imports everywhere?

**Comment by Emantor** [comment]
Relative paths are used since the imports key is usually used to import files for the local test definition which should live in the same repository. Our usecase for this is usually a customer and board specific strategy for the target hardware. For global additions to labgrid I would recommend not using the imports key and rather use your own python module installed into the same virtualenv.

**Comment by sjg20** [comment]
@Emantor What is the 'import key'? I had thought that Python encourages absolute imports. I have tended to uses absolute imports in my own Python programs and have not hit any problems with that

**Reply by jluebbe** [reply]
I find the relative imports (with `.`) useful as a clear distinction to external dependencies. Do you have a link where absolute imports for internal dependencies are encouraged?

**Comment by Emantor** [comment]
The device configuration YAML contains an [imports](https://labgrid.readthedocs.io/en/latest/man/device-config.html#imports) key which can be a list of python files that are imported before target creation.

**Reply by jluebbe** [reply]
I think @sjg20 is referring to the import statements in python modules, not the environment yaml.

**Comment by sjg20** [comment]
Oh yes, I am referring to the 'import' statements.

In Chromium OS we require absolute imports since it makes it easier to see where the module is coming from and avoids needing to add . and .. plus subdirs (depending on where your module is) with lots of trial and error. I have definitely found this annoying with labgrid.

As mentioned I used to use relative imports for the Python tools I wrote (patman, buildman, binman) but switched to absolute to make it easier to share modules across packages.

I see a mention of this in the tutorial video:

https://www.youtube.com/watch?v=APIA8aUH98Q (at 1:10 mark)

There is much discussion about relative vs. absolute...e.g. this one:

https://chrisyeh96.github.io/2017/08/08/definitive-guide-python-imports.html#absolute-vs-relative-import

**Reply by jluebbe** [reply]
Very widely used and modern Python packages use explicit relative imports:
https://github.com/psf/requests/blob/main/src/requests/__init__.py#L147
https://github.com/pypa/build/blob/main/src/build/__main__.py#L24
https://github.com/urllib3/urllib3/blob/main/src/urllib3/__init__.py#L14
https://github.com/pypa/setuptools/blob/main/setuptools/__init__.py#L13
https://github.com/numpy/numpy/blob/main/numpy/__init__.py#L92

So although I can see why some prefer absolute imports, I'd prefer to stay with the widely used convention of explicit relative imports.

**Comment by sjg20** [comment]
Er, really?

https://github.com/numpy/numpy/blob/main/numpy/lib/_arraysetops_impl.py#L23
https://github.com/numpy/numpy/blob/main/numpy/lib/_array_utils_impl.py

Anyway, thank you for considering it. I will fumble along with my trial-and-error approach to imports :-)
